### ISSUE TITLE
Consider a simpler interface for connecting Components

### ISSUE THREAD

**Issue by tkuchida** [opensim-org/opensim-core#1118]
For example, a user will be thinking "I want to connect this Output to my ConsoleReporter" but must reinterpret this action as "I need to update the Input of my ConsoleReporter to connect to this Output":
```cpp
myReporter->updInput("inputs").connect(myModel.getOutput(pathToOutput), annotation);
```
A more intuitive interface might be to hide the directionality of the connection, which may not have meaning for the user. For example:
```cpp
myModel.connect(pathToInput, pathToOutput, annotation);
myModel.connect(pathToOutput, pathToInput, annotation); //equivalent
```

**Comment by chrisdembia** [comment]
Related #508.

**Comment by chrisdembia** [comment]
I'd prefer putting the function the OpenSim namespace as a free function: `OpenSim::connect(output, input)`.

**Comment by tkuchida** [comment]
> I'd prefer putting the function the OpenSim namespace as a free function: `OpenSim::connect(output, input)`.

Seems reasonable if the arguments are references, but if `connect()` is a member function of Component, then the arguments could potentially also be paths relative to that Component. Alternatively, we could just rely on `getComponent()`'s path-handling capabilities to return references for `connect()`, as in
```cpp
OpenSim::connect(myModel.getComponent(pathToOutput), myModel.getComponent(pathToInput));
```

**Comment by tkuchida** [comment]
### Use case
In exampleHopperDevice.cpp (Step 3, Task B):
```cpp
kneeDevice->updComponent("controller").updInput("activation")
    .connect(assistedHopper.getComponent(vastus).getOutput("activation"));
```
is connecting an output to an input, but the word "connect" appears in the *middle* of the line. Perhaps
```cpp
connect( assistedHopper.getComponent(vastus).getOutput("activation"),
         kneeDevice->updComponent("controller").updInput("activation") );
```
would be better because "connect" appears at the beginning.

**Comment by chrisdembia** [comment]
Your proposal is very similar to how [Qt's signals and slots](http://doc.qt.io/qt-4.8/signalsandslots.html) work:

> `connect(Object1, signal1, Object2, slot2)`

So I think it should work!

The main issue I forsee is that Java does not support free functions--only static class methods. Python does allow free functions, so we can achieve the same syntax in Python and C++.

**Comment by tkuchida** [comment]
> (not sure...someone should check) which may be confusing to users

I've added the "Needs Discussion" label. We should all agree on what we want before starting to implement (see #1116).

**Comment by chrisdembia** [comment]
If the `connect(Obj1, signal1, Obj2, slot2)` syntax is possible in MATLAB, then I support this syntax.

**Comment by tkuchida** [comment]
> If the `connect(Obj1, signal1, Obj2, slot2)` syntax is possible in MATLAB

@aymanhab Do you know whether it's possible to swig an `OpenSim::connect()` method into Matlab?

**Comment by aymanhab** [comment]
Free standing functions do not exist in Java but we can hack some workarounds,  
 1.  if we leave it to swig it becomes a static method of the corresponding module class (and look something like opensimSimulationJNI.connect( , , ,  ) 
 2.  we can add a class called "connect" (in Java Only)  and have it delegate the call, but usually class names are nouns not verbs.
 3. It'd feel cleaner and more extensible making a class, call it ConnectionMaker or ConnectionManager and make it have a connect method and few other support functions (e.g. checkConnection, dumpConnection, ....)

**Comment by chrisdembia** [comment]
I feel that a class like ConnectionMaker, etc. would be confusing to users, but you make a good point about perhaps wanting other related methods. I think the model should just have the methods then:

```c++
model.connect(output, input);
```

**Comment by aseth1** [comment]
The `Component` is the connection maker and the manager of its connections,  thus adding another Object will only obfuscate that fact.  I like the convenience of the proposed interfaces and we could free  up `Component::connect()` to do this if we renamed the current `Commponent::connect()` (maybe to `finalizeConnections()` in keeping with `finalizeFromProperties()`), since it confusing with the act of making a single connection.

Then we could overload c `Component::connect()` with:
 `connect (Input& , const Output& out)`  and  `connect (const Output& out, Input&  in)` with the caveat/check that the Input->Output case, the Input is a member of the invoked Component and in the Output->Input case the Output is a member of the invoked Component for the sake of clarity. We could allow Component C connect A and B, but that is more confusing  to me. At least you can use what ever order suits your predilection, but under the covers it would be identical (`in.connect(out)`)

**Comment by aseth1** [comment]
I tend to think both examples in the *Use case* above equally confusing by virtue of their length.
Perhaps this would be easier to understand?

```cpp
// get a modifiable reference to the controller's activation input
auto& controllerInput = kneeDevice->updComponent("controller").updInput("activation");
// get the vastus muscle's activation output
auto& muscleOutput = assistedHopper.getComponent(vastus).getOutput("activation"));
// connect the controller's input to the muscle's output
controllerInput.connect(muscleOutput);
```

**Comment by tkuchida** [comment]
I like `Component::connect(Output, Input)`. We could enforce not only that the Input is a member of the Component, but also that the Output is a descendant (if, philosophically, we allow a Component to "see" only its descendants); however, if the philosophy is moving towards UUIDs (#1192), then connecting to arbitrary Outputs seems fine.

> I tend to think both examples in the *Use case* above equally confusing by virtue of their length.

The first example in the *Use case* above is the current design, *not* the proposal. The second example could be broken up as follows:
```cpp
auto& vastusOutput = assistedHopper.getComponent(vastus).getOutput("activation");
auto& controllerInput = kneeDevice->updComponent("controller").updInput("activation");
//connect(vastusActivationOutput, controllerInput);
myModel.connect(vastusActivationOutput, controllerInput);
```
The `controllerInput.connect(muscleOutput);` still requires the user to think about satisfying inputs rather than grabbing an available output wire and plugging it into an input port (the latter is more natural to me at least). I would be fine with the new proposal if it were also possible to do `vastusActivationOutput.connect(controllerInput);`.

**Comment by aseth1** [comment]
Currently, you can connect to any Output that is part of your tree not just your descendants.

>> I tend to think both examples in the Use case above equally confusing by virtue of their length.

>The first example in the Use case above is the current design, not the proposal. The second example could be broken up as follows:

I meant **both** the existing and the proposal are both equally confusing by virtue of their length. Yes, breaking them up improves them in both cases as well. That is my point. Breaking it up makes it readable in both cases.

We should able to overload `connect()` to permit both orders to meet different preferences as indicated. Are you advocating only having Output->Input order and not Input->Output?

**Comment by tkuchida** [comment]
> Yes, breaking them up improves them in both cases as well. That is my point. Breaking it up makes it readable in both cases.

The issue is not with readability or parsing, it is that we are unnecessarily constraining the way the user can think about model building. The comment at the top of the issue describes the problem I'm trying to solve: as a user, it is more natural for me to think "I want to connect this Output to my ConsoleReporter" rather than "I want to update the Input of my ConsoleReporter to connect to this Output". The symmetrical `connect(input|output, output|input)` was appealing for this reason.

> Are you advocating only having Output->Input order and not Input->Output?

Sorry for my imprecision. My reply was intended to indicate that I like the idea of having an "omniscient" `connect()` method that isn't tied to (a member function of) either the input or the output. My mental model is that there are Components sitting on the table and I'm wiring them together (where I'm OpenSim or some ancestor of the Components being wired together). Removing the directionality of forming connections is the objective of this issue.

**Comment by aymanhab** [comment]
I tend to agree with @tkuchida s mental picture that connection is not done by a component (asymmetric) but by a "higher level authority" that connects two wires. Earlier I mentioned ConnectionManager, but that could also be a static class method of Component, eitherway that fits the mental picture more than an Object.method paradigm as far as I can tell.

**Comment by aseth1** [comment]
> We should able to overload connect() to permit both orders to meet different preferences as indicated

I think the two connect() variants could also be static methods of Component instead of overloaded member functions.

**Comment by jenhicks** [comment]
Closed by Dev Prop #1219
